File: Golang/746.min-cost-climbing-stairs.go
/*
 * @lc app=leetcode.cn id=746 lang=go
 *
 * [746] 使用最小花费爬楼梯
 */
package leetcode

// @lc code=start
func minCostClimbingStairs(cost []int) int {
	var stepStamp int = len(cost)
	var tempVec = map[int]int{
		0: cost[0],
		1: cost[1],
	}
	var stepCost int = 0
	for idx := 2; idx <= stepStamp; idx += 1 {
		if idx == stepStamp {
			stepCost = 0
		} else {
			stepCost = cost[idx]
		}
		// golang 1.21 已经支持max, min函数
		tempVec[idx%2] = min(tempVec[0], tempVec[1]) + stepCost
	}

	return tempVec[stepStamp%2]
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

// @lc code=end

File: Cplusplus/746.min-cost-climbing-stairs.cpp
/*
 * @lc app=leetcode.cn id=746 lang=cpp
 *
 * [746] 使用最小花费爬楼梯
 */
#include <vector>
using namespace std;

// @lc code=start
class Solution
{
public:
    int minCostClimbingStairs(vector<int> &cost)
    {
        // 可以从0 或者 1号台阶开始
        // 每一次支付完成之后，可以选择爬 1 或 2个台阶
        int stepStamp = cost.size();
        // set
        vector<int> tempVec = vector<int>(2, 0);
        tempVec[0] = cost[0];
        tempVec[1] = cost[1];
        // regressive
        cost.push_back(0); // 最后楼顶台阶，费用为0
        for (int idx = 2; idx <= stepStamp; ++idx)
        {
            tempVec[idx % 2] = min(tempVec[0], tempVec[1]) + cost[idx];
        }

        return tempVec[(stepStamp) % 2];
    }
};
// @lc code=end
